authorization/token/service/token_service_blackbox_test.go

```diff
@@ -0,0 +1,493 @@
+	tm := testtoken.TokenManager
+
+	// Create a user
+	u := s.Graph.CreateUser()
+
+	// Create an access token for the user
+	at, err := tm.GenerateUserTokenForIdentity(s.Ctx, *u.Identity(), false)
+	require.NoError(s.T(), err)
+
+	// Create a new resource type, with scope "echo"
+	rt := s.Graph.CreateResourceType()
+	rt.AddScope("echo")
+
+	// Create a role with scope echo
+	echoRole := s.Graph.CreateRole(rt)
+	echoRole.AddScope("echo")
+
+	// Create a resource with the new resource type
+	r := s.Graph.CreateResource(rt)
+
+	// Assign the echo role to the user for the new resource
+	s.Graph.CreateIdentityRole(u, r, echoRole)
+
+	// Audit the user token for the new resource ID
+	rptToken, err := s.Application.TokenService().Audit(tokencontext.ContextWithTokenManager(s.Ctx, tm), u.Identity(), at.AccessToken, r.ResourceID())
+	require.NoError(s.T(), err)
+
+	require.NotNil(s.T(), rptToken)
+
+	// Parse the signed RPT token
+	tokenClaims, err := tm.ParseToken(s.Ctx, *rptToken)
+	require.NoError(s.T(), err)
+
+	// There should be one permission in the token, and it should be for the resource
+	perms := *tokenClaims.Permissions
+	require.Len(s.T(), perms, 1)
+	require.Equal(s.T(), *perms[0].ResourceSetID, r.ResourceID())
+	// There should be one scope, "echo"
+	require.Len(s.T(), perms[0].Scopes, 1)
+	require.Contains(s.T(), perms[0].Scopes, "echo")
+
+	// Audit the RPT token for the same resource ID, it should return nil since the privileges haven't changed and the token should not have expired
+	rptToken, err = s.Application.TokenService().Audit(tokencontext.ContextWithTokenManager(s.Ctx, tm), u.Identity(), *rptToken, r.ResourceID())
+	require.NoError(s.T(), err)
+	require.Nil(s.T(), rptToken)
+}
+
+func (s *tokenServiceBlackboxTest) TestRPTTokenReplacedWithAdditionalResource() {
+	tm := testtoken.TokenManager
+
+	// Create a user
+	u := s.Graph.CreateUser()
+
+	// Create an access token for the user
+	at, err := tm.GenerateUserTokenForIdentity(s.Ctx, *u.Identity(), false)
+	require.NoError(s.T(), err)
+
+	// Create a new resource type, with scopes "foxtrot" and "golf"
+	rt := s.Graph.CreateResourceType()
+	rt.AddScope("foxtrot")
+	rt.AddScope("golf")
+
+	// Create a role with scope foxtrot
+	foxtrotRole := s.Graph.CreateRole(rt)
+	foxtrotRole.AddScope("foxtrot")
+
+	// Create a role with scope golf
+	golfRole := s.Graph.CreateRole(rt)
+	golfRole.AddScope("golf")
+
+	// Create a resource with the new resource type
+	r := s.Graph.CreateResource(rt)
+
+	// Assign the foxtrot role to the user for the new resource
+	s.Graph.CreateIdentityRole(u, r, foxtrotRole)
+
+	// Audit the user token for the new resource ID
+	rptToken, err := s.Application.TokenService().Audit(tokencontext.ContextWithTokenManager(s.Ctx, tm), u.Identity(), at.AccessToken, r.ResourceID())
+	require.NoError(s.T(), err)
+
+	require.NotNil(s.T(), rptToken)
+
+	// Parse the signed RPT token
+	tokenClaims, err := tm.ParseToken(s.Ctx, *rptToken)
```

Comment:
@sbryzak , Could you add a small token validation test ( add a function and reuse it for all ) to ensure the typical token claims exists with the right values? This is to ensure - that since are exchanging a normal user access token with an RPT - the usual token claims like `preferred_username` , `sub`, `email` should remain as is.

Reply:
Good idea, I'll add this to the token tests.

Reply:
Done